File: src/com/avail/environment/editor/utility/PrefixNode.java
package com.avail.environment.editor.utility;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeSet;
import java.util.function.Consumer;

import static java.util.Collections.unmodifiableList;

/**
 * A {@code PrefixNode} is a node in a {@link PrefixTrie}.
 *
 * @param <T>
 *        The type of object held at the {@code PrefixNode}.
 * @author Rich Arriaga &lt;rich@availlang.org&gt;
 */
public class PrefixNode<T>
{
	/**
	 * The {@link Map} that holds the next nodes in the trie.
	 */
	private final Map<Character, PrefixNode<T>> children =
		new HashMap<>();

	/**
	 * The target content of this node if any exist, otherwise {@code null}.
	 */
	private @Nullable T content;

	/**
	 * The height position in the {@link PrefixTrie}.
	 */
	private final int depth;

	/**
	 * An object to lock editing to this {@link PrefixNode}.
	 */
	private final Object lock = new Object();

	/**
	 * Answer the {@link #content}.
	 *
	 * @return A String.
	 */
	public @Nullable T content ()
	{
		return content;
	}

	/**
	 * The list of viable complete prefix word from this point in the
	 * {@link PrefixTrie}.
	 */
	private List<String> words = new ArrayList<>();

	/**
	 * Add a word to this {@code PrefixNode}.
	 *
	 * @param word
	 *        The word to add.
	 */
	private void addWord (final String word)
	{
		synchronized (lock)
		{
			final TreeSet<String> set = new TreeSet<>(words);
			set.add(word);
			words = new ArrayList<>(set);
		}
	}

	/**
	 * Answer the {@link #words} from this {@code PrefixNode}.
	 *
	 * @return A {@link List}.
	 */
	public final List<String> words ()
	{
		synchronized (lock)
		{
			return unmodifiableList(words);
		}
	}

	/**
	 * Add a word to this {@code PrefixNode}.
	 *
	 * @param word
	 *        The word branch to add.
	 * @param contents
	 *        The target content of the target node if any exist, otherwise
	 *        {@code null}.
	 */
	public void addWord (
		final String word,
		final T contents)
	{

		addWord(word);

		if (word.length() - 1 > depth)
		{
			final PrefixNode<T> nextNode;
			synchronized (lock)
			{
				nextNode = children.computeIfAbsent(
					word.charAt(depth + 1),
					key -> new PrefixNode<>(depth + 1));
			}
			nextNode.addWord(word, contents);
		}
		else
		{
			this.content = contents;
		}
	}

	/**
	 * Search the {@link PrefixTrie} starting at this {@code PrefixNode}.
	 *
	 * @param word
	 *        The word to search.
	 */
	public @Nullable PrefixNode<T> searchTrie (
		final String word)
	{
		if (word.length() - 1 > depth)
		{
			final PrefixNode<T> nextNode = nextNodeFor(word.charAt(depth + 1));
			return nextNode != null
				? nextNode.searchTrie(word)
				: null;
		}
		else
		{
			return this;
		}
	}

	/**
	 * Search the {@link PrefixTrie} starting at this {@code PrefixNode}. Add
	 * visited nodes to the consumer.
	 *
	 * @param word
	 *        The word to search.
	 * @param acceptEachNode
	 *        A {@link Consumer} that accepts a {@code PrefixNode}.
	 */
	public @Nullable PrefixNode<T> searchTrie (
		final String word,
		final Consumer<PrefixNode<T>> acceptEachNode)
	{
		if (word.length() - 1 > depth)
		{
			final PrefixNode<T> nextNode = nextNodeFor(word.charAt(depth + 1));
			acceptEachNode.accept(nextNode);
			return nextNode != null
				? nextNode.searchTrie(word, acceptEachNode)
				: null;
		}
		else
		{
			acceptEachNode.accept(this);
			return this;
		}
	}

	/**
	 * Answer the next {@code PrefixNode} for the given character.
	 *
	 * @param c
	 *        The char to get.
	 * @return A {@code PrefixNode} if it exists; {@code null} otherwise.
	 */
	public @Nullable PrefixNode<T> nextNodeFor (final char c)
	{
		return children.getOrDefault(c, null);
	}

	/**
	 * Construct an empty {@code PrefixNode}.
	 *
	 * @param depth
	 *        Record its depth in the trie.
	 */
	public PrefixNode (
		final int depth)
	{
		this.depth = depth;
	}
}

File: avail-workbench/src/main/kotlin/com/avail/environment/editor/utility/PrefixNode.kt
package com.avail.environment.editor.utility

/**
 * A `PrefixNode` is a node in a [PrefixTrie].
 *
 * @param T
 *   The type of object held at the `PrefixNode`.
 * @property depth
 *   Record its depth in the trie.
 *
 * @author Rich Arriaga &lt;rich@availlang.org&gt;
 *
 * @constructor
 * Construct an empty `PrefixNode`.
 *
 * @param depth
 *   Record its depth in the trie.
 */
class PrefixNode<T> constructor(private val depth: Int)
{
	/**
	 * The [Map] that holds the next nodes in the trie.
	 */
	private val children = mutableMapOf<Char, PrefixNode<T>>()

	/**
	 * The target content of this node if any exist, otherwise `null`.
	 */
	private var content: T? = null

	/**
	 * An object to lock editing to this [PrefixNode].
	 */
	private val lock = Any()

	/**
	 * The list of viable complete prefix word from this point in the
	 * [PrefixTrie].
	 */
	val words = mutableSetOf<String>()

	/**
	 * Answer the [content].
	 *
	 * @return A String.
	 */
	fun content(): T?
	{
		return content
	}

	/**
	 * Add a word to this `PrefixNode`.
	 *
	 * @param word
	 *   The word to add.
	 */
	private fun addWord(word: String)
	{
		synchronized(lock) { words.add(word) }
	}

	/**
	 * Answer the [words] from this `PrefixNode`.
	 *
	 * @return A [Set].
	 */
	fun words(): Set<String>
	{
		synchronized(lock) {
			return words.toSet()
		}
	}

	/**
	 * Add a word to this `PrefixNode`.
	 *
	 * @param word
	 *   The word branch to add.
	 * @param contents
	 *   The target content of the target node if any exist, otherwise `null`.
	 */
	fun addWord(word: String, contents: T)
	{
		addWord(word)
		if (word.length - 1 > depth)
		{
			val nextNode: PrefixNode<T>
			synchronized(lock)
			{
				nextNode = (children).computeIfAbsent(word[depth + 1])
					{ PrefixNode(depth + 1) }
			}
			nextNode.addWord(word, contents)
		}
		else
		{
			this.content = contents
		}
	}

	/**
	 * Search the [PrefixTrie] starting at this `PrefixNode`.
	 *
	 * @param word
	 *   The word to search.
	 */
	fun searchTrie(word: String): PrefixNode<T>? =
		if (word.length - 1 > depth)
		{
			val nextNode = nextNodeFor(word[depth + 1])
			nextNode?.searchTrie(word)
		}
		else
		{
			this
		}

	/**
	 * Search the [PrefixTrie] starting at this `PrefixNode`. Add visited nodes
	 * to the consumer.
	 *
	 * @param word
	 *   The word to search.
	 * @param acceptEachNode
	 *   A lambda that accepts a `PrefixNode`.
	 */
	fun searchTrie(word: String, acceptEachNode: (PrefixNode<T>?) -> Unit)
		: PrefixNode<T>? =
			if (word.length - 1 > depth)
			{
				val nextNode = nextNodeFor(word[depth + 1])
				acceptEachNode.invoke(nextNode)
				nextNode?.searchTrie(word, acceptEachNode)
			}
			else
			{
				acceptEachNode.invoke(this)
				this
			}

	/**
	 * Answer the next `PrefixNode` for the given character.
	 *
	 * @param c
	 *   The char to get.
	 * @return A `PrefixNode` if it exists; `null` otherwise.
	 */
	fun nextNodeFor(c: Char): PrefixNode<T>?
	{
		return children.getOrDefault(c, null)
	}
}
